stan/math/prim/functor/apply_scalar_ternary.hpp

```diff
@@ -0,0 +1,220 @@
+#ifndef STAN_MATH_PRIM_FUNCTOR_APPLY_SCALAR_TERNARY_HPP
+#define STAN_MATH_PRIM_FUNCTOR_APPLY_SCALAR_TERNARY_HPP
+
+#include <stan/math/prim/fun/as_column_vector_or_scalar.hpp>
+#include <stan/math/prim/meta/is_stan_scalar.hpp>
+#include <stan/math/prim/meta/is_container.hpp>
+#include <stan/math/prim/meta/is_eigen.hpp>
+#include <stan/math/prim/meta/require_generics.hpp>
+#include <stan/math/prim/err/check_matching_dims.hpp>
+#include <stan/math/prim/err/check_matching_sizes.hpp>
+#include <stan/math/prim/fun/num_elements.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Base template function for vectorization of ternary scalar functions
+ * defined by applying a functor to a combination of scalars,
+ * containers of matching sizes, or a combination of a scalar and a container.
+ * These containers can be a standard library vector, Eigen dense
+ * matrix expression template, or container of these. For each specialisation,
+ * the same type as the largest (dimension) input is returned.
+ *
+ * This base template function takes (and returns) scalars.
+ *
+ * @tparam T1 An Arithmetic, Integral, var, or fvar<T> type.
+ * @tparam T2 An Arithmetic, Integral, var, or fvar<T> type.
+ * @tparam T3 An Arithmetic, Integral, var, or fvar<T> type.
+ * @tparam F Type of functor to apply.
+ * @param x First input to which operation is applied.
+ * @param y Second input to which operation is applied.
+ * @param z Third scalar to which operation is applied.
+ * @param f functor to apply to inputs.
+ * @return Scalar with result of applying functor to inputs.
+ */
+template <typename F, typename T1, typename T2, typename T3,
+          require_all_stan_scalar_t<T1, T2, T3>* = nullptr>
+inline auto apply_scalar_ternary(const F& f, const T1& x, const T2& y,
+                                 const T3& z) {
+  return f(x, y, z);
+}
+
+/**
+ * Specialisation for use with three Eigen inputs. Eigen's ternaryExpr framework
+ * is used for more efficient indexing of both row- and column-major inputs
+ * without separate loops.
+ *
+ * @tparam T1 Eigen type of first argument to which functor is applied.
+ * @tparam T2 Eigen type Type of second argument to which functor is applied.
+ * @tparam T3 Eigen type of third argument to which functor is applied.
+ * @tparam F Type of functor to apply.
+ * @param x First Eigen input to which operation is applied.
+ * @param y Second Eigen input to which operation is applied.
+ * @param z Third Eigen input to which operation is applied.
+ * @param f functor to apply to Eigen input.
+ * @return Eigen object with result of applying functor to inputs.
+ */
+template <typename F, typename T1, typename T2, typename T3,
+          require_all_eigen_t<T1, T2, T3>* = nullptr>
+inline auto apply_scalar_ternary(F&& f, T1&& x, T2&& y, T3&& z) {
+  check_matching_dims("Ternary function", "x", x, "y", y);
+  check_matching_dims("Ternary function", "y", y, "z", z);
+  check_matching_dims("Ternary function", "x", x, "z", z);
```

Comment:
[optional] if `x` and `y` have matching dims and `y` and `z` have matching dims then `x` and `z` have matching dims. This could change in a few places